llvm/lib/Transforms/Utils/LoopIterCountsToAssumptions.cpp

```diff
@@ -0,0 +1,201 @@
+    return nullptr;
+  }
+  return LoopCmpInstr;
+}
+
+} // namespace
+
+PreservedAnalyses
+LoopIterCountsToAssumptions::run(Loop &CurrentLoop, LoopAnalysisManager &AM,
+                                 LoopStandardAnalysisResults &AR,
+                                 LPMUpdater &U) {
+  ScalarEvolution *SE = &AR.SE;
+  AssumptionCache *AC = &AR.AC;
+  BasicBlock *LoopHeader = CurrentLoop.getHeader();
+  Function *F = LoopHeader->getParent();
+
+  std::optional<int64_t> RawMinIterationCount = getMinTripCount(&CurrentLoop);
+
+  // dump loop summary
+  LLVM_DEBUG(if (CurrentLoop.getLoopPreheader()) {
+    dbgs() << "Preheader:" << CurrentLoop.getLoopPreheader()->getName() << "\n";
+  });
+  LLVM_DEBUG(dbgs() << "Header:" << LoopHeader->getName() << "\n");
+
+  if (!RawMinIterationCount.has_value() || RawMinIterationCount.value() <= 0)
+    return PreservedAnalyses::all();
+  int64_t MinIterCount = RawMinIterationCount.value();
+
+  LLVM_DEBUG(dbgs() << "Processing Loop Iteration Count Metadata: "
+                    << F->getName() << " " << CurrentLoop.getName() << " ("
+                    << MinIterCount << ")\n");
+
+  ICmpInst *LoopCmpInstr = getLoopCmpInst(CurrentLoop);
+  if (!LoopCmpInstr)
+    return PreservedAnalyses::all();
+
+  // get Scalar Evolution of the induction variable
+  std::pair<const SCEVAddRecExpr *, int> BundledAddRecExpr =
+      getLoopCmpIVAddRecExpr(*LoopCmpInstr, SE);
+  const SCEVAddRecExpr *AddRecExpr = BundledAddRecExpr.first;
+  int IVOperand = BundledAddRecExpr.second;
+  if (!AddRecExpr) {
+    LLVM_DEBUG(
+        dbgs() << "LoopIterCountsToAssumptions-Warning: Could not extract "
+                  "SCEVAddRecExpr! Will not process Loop Iteration Count\n");
+    return PreservedAnalyses::all();
+  }
+
+  LLVM_DEBUG(dbgs() << "Loop Iteration Count        : " << MinIterCount
+                    << "\n");
+
+  Instruction *HeaderTerminator = LoopHeader->getTerminator();
+  IRBuilder<> Builder(HeaderTerminator);
+
+  SCEVExpander Expander(*SE, F->getParent()->getDataLayout(), "expanded");
+
+  // get Minimum Iteration Value Expression
+  const SCEV *MinIterSCEV = AddRecExpr->evaluateAtIteration(
+      SE->getConstant(APInt(32, MinIterCount)), *SE);
+
+  // in the case of decrementing, add NSW flag, so that the expression can be
+  // simplified
+  if (auto *AE = dyn_cast<SCEVAddExpr>(MinIterSCEV)) {
+    SmallVector<const SCEV *> Ops(AE->operands());
+    MinIterSCEV = SE->getAddExpr(Ops, SCEV::FlagNSW);
+  }
+
+  Type *Op0Type = LoopCmpInstr->getOperand(0)->getType();
+  Value *MinIterValue =
+      Expander.expandCodeFor(MinIterSCEV, Op0Type, HeaderTerminator);
+
+  LLVM_DEBUG(dbgs() << "MinIterValue: "; MinIterValue->dump());
+
+  Value *LoopStartValue =
+      Expander.expandCodeFor(AddRecExpr->getStart(), Op0Type, HeaderTerminator);
+
+  // if the loop is decrementing, the Initial Start of the loop is already in
+  // the MinIterValue, therefore it does not have to be added again
```

Comment:
I'm not sure I understand what we are doing here. The comment mentions decrementing, but this is not what is checked in `containStartValue`

Reply:
removed that check, since I was not actually using it